presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1365,35 +1380,39 @@ private VersionType toVersionType(TableVersionType type)
                 case VERSION:
                     return VersionType.VERSION;
             }
-            throw new SemanticException(NOT_SUPPORTED, type.toString(), "Table version type not supported.");
+            throw new SemanticException(NOT_SUPPORTED, "Table version type %s not supported." + type);
         }
         private Optional<TableHandle> processTableVersion(Table table, QualifiedObjectName name, Optional<Scope> scope)
         {
-            Expression asOfExpr = table.getTableVersionExpression().get().getAsOfExpression();
+            Expression stateExpr = table.getTableVersionExpression().get().getStateExpression();
             TableVersionType tableVersionType = table.getTableVersionExpression().get().getTableVersionType();
-            ExpressionAnalysis expressionAnalysis = analyzeExpression(asOfExpr, scope.get());
+            TableVersionOperator tableVersionOperator = table.getTableVersionExpression().get().getTableVersionOperator();
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(stateExpr, scope.get());
             analysis.recordSubqueries(table, expressionAnalysis);
-            Type asOfExprType = expressionAnalysis.getType(asOfExpr);
-            if (asOfExprType == UNKNOWN) {
-                throw new PrestoException(INVALID_ARGUMENTS, format("Table version AS OF expression cannot be NULL for %s", name.toString()));
+            Type stateExprType = expressionAnalysis.getType(stateExpr);
+            if (tableVersionOperator == LESS_THAN) {
+                throw new PrestoException(INVALID_ARGUMENTS, format("Table version BEFORE expression is not supported for %s", name.toString()));
+            }
+            if (stateExprType == UNKNOWN) {
+                throw new PrestoException(INVALID_ARGUMENTS, format("Table version AS OF/BEFORE expression cannot be NULL for %s", name.toString()));
             }
-            Object evalAsOfExpr = evaluateConstantExpression(asOfExpr, asOfExprType, metadata, session, analysis.getParameters());
+            Object evalStateExpr = evaluateConstantExpression(stateExpr, stateExprType, metadata, session, analysis.getParameters());
             if (tableVersionType == TIMESTAMP) {
-                if (!(asOfExprType instanceof TimestampWithTimeZoneType)) {
-                    throw new SemanticException(TYPE_MISMATCH, asOfExpr,
+                if (!(stateExprType instanceof TimestampWithTimeZoneType)) {
+                    throw new SemanticException(TYPE_MISMATCH, stateExpr,
                             "Type %s is invalid. Supported table version AS OF expression type is Timestamp with Time Zone.",
```

Comment:
Update error message here and the one below

Reply:
Done